lib/Dialect/MSFT/ExportQuartusTcl.cpp

```diff
@@ -73,34 +91,71 @@ static void emit(TclOutputState &s, PlacementDB::PlacedInstance inst,
 
   // To which entity does this apply?
   s.os << " -to $parent|";
-  emitPath(s, inst.path);
+  emitPath(s, inst.path, symCache);
   // If instance name is specified, add it in between the parent entity path and
-  // the child entity patch.
-  if (auto instOp = dyn_cast<hw::InstanceOp>(inst.op))
+  // the child entity path.
+  if (auto instOp = dyn_cast<msft::InstanceOp>(inst.op))
     s.os << instOp.instanceName() << '|';
   else if (auto name = inst.op->getAttrOfType<StringAttr>("name"))
     s.os << name.getValue() << '|';
   s.os << inst.subpath << '\n';
 }
 
+/// Create a SymbolCache to use during Tcl export.
+void circt::msft::populateSymbolCache(mlir::ModuleOp mod, SymbolCache &cache) {
+  // Traverse each module and each instance within the module.
+  for (auto msftMod : mod.getOps<MSFTModuleOp>()) {
+    for (auto inst : msftMod.getOps<msft::InstanceOp>()) {
+      // Use the instance symbol name.
+      StringAttr symName = inst.sym_nameAttr();
+
+      // Add the symbol to the cache.
+      cache.addDefinition(symName, inst);
+    }
+  }
+
+  cache.freeze();
+}
+
 /// Write out all the relevant tcl commands. Create one 'proc' per module which
 /// takes the parent entity name since we don't assume that the created module
 /// is the top level for the entire design.
-LogicalResult circt::msft::exportQuartusTcl(hw::HWModuleOp hwMod,
-                                            llvm::raw_ostream &os) {
+LogicalResult circt::msft::exportQuartusTcl(MSFTModuleOp hwMod,
+                                            SymbolCache &symCache,
+                                            StringRef outputFile) {
+  // Build up the output Tcl, tracking symbol references in state.
+  std::string s;
+  llvm::raw_string_ostream os(s);
   TclOutputState state(os);
   PlacementDB db(hwMod);
   size_t failures = db.addDesignPlacements();
   if (failures != 0)
     return hwMod->emitError("Could not place ") << failures << " instances";
 
+  // Don't emit empty procs if the database is empty.
+  if (db.empty())
```

Comment:
I don't think we want to skip emitting a proc which the user has requested be generated. Other tcl may depend on it being there, even if it is empty.

Reply:
Ah, makes sense. I will roll it into the change to accept a list of top modules to generate Tcl for.